third_party/nvfuser/csrc/fusion_segmenter.cpp

```diff
@@ -1545,24 +1694,128 @@ SegmentedGroup* SegmentCandidateFinder::mergeAllGivenGroups(
 }
 namespace {
 
-// Guard to temporarily change the inputs and outputs of a fusion. On
+// Guard to temporarily change the inputs and outputs of a
+// fusion. Cast expressions to fp32 and fp16 are also inserted. On
 // destruction will return fusion to original state.
-// Not used temporarily but will be useful when adding more mergin heuristics
 class FusionSegmentGuard : public NonCopyable {
  public:
   FusionSegmentGuard() = delete;
 
+  // Narrow the fusion to a segment defined by inputs and outputs
```

Comment:
So looks like you have four options:
Narrow only
Cast Only
Merge groups, cast, and narrow
Cast and narrow

Reply:
Yeah, wanted to simplify, but looks like all of these variations are necessary (unless significantly refactored)